### PR TITLE
Fix inaccurate NY supplemental tax calculations

### LINKED ISSUE #1679: Fix inaccurate NY supplemental tax calculations
In some cases, the 0.200.6 version of PolicyEngineUS produces incorrect 2021 values for the `ny_supplemental_tax` variable, and therefore, incorrect 2021 values for the `ny_income_tax` variable.

Currently there are no unit tests for the `ny_supplemental_tax` variable formula.
Below I show two units tests: the first of which passes and the second of which fails.

```
- name: Unit test 1 for ny_supplemental_tax
  absolute_error_margin: 0.01
  period: 2021
  input:
    state_code: NY
    filing_status: JOINT
    ny_agi: 256_000
    ny_taxable_income: 238_950
    ny_main_income_tax: 14_069.755
  output:
    ny_supplemental_tax: 1_055.78
    # calculated by hand using input and worksheet 2 on page 52 of the
    # 2021 NY Form IT-201 Instructions, which is at the following URL:
    # https://www.tax.ny.gov/pdf/2021/inc/it201i_2021.pdf#page=52

- name: Unit test 2 for ny_supplemental_tax
  absolute_error_margin: 0.01
  period: 2021
  input:
    state_code: NY
    filing_status: JOINT
    ny_agi: 324_030
    ny_taxable_income: 307_980
    ny_main_income_tax: 18_439.35
  output:
    ny_supplemental_tax: 1_055.78
    # calculated by hand using input and worksheet 2 on page 52 of the
    # 2021 NY Form IT-201 Instructions, which is at the following URL:
    # https://www.tax.ny.gov/pdf/2021/inc/it201i_2021.pdf#page=52
```

The second unit test fails with an output difference of $1,038.25.

### PR DESCRIPTION
**Pull request by Shinuing**
Fixes #1679

### Summary
🐛📝🚀

This pull request updates the New York state income tax calculations for the policy engine, adding the latest parameters, references, and tests. It fixes the ny_supplemental_tax variable and adds new files for the incremental benefit and recapture base amounts for each filing status. It also bumps the patch version and updates the changelog.

> _Oh we're the crew of the `ny_supplemental_tax`_
> _We fix the brackets and the rates with every pull request_
> _We heave and ho and test the code, we make it shipshape and exact_
> _And we bump the patch version when we're done with the `ny_supplemental_tax`_

### Walkthrough
*  Bump patch version and fix New York supplemental tax calculation ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-5885042762833a4aef51405a4036067e4e6aefa7a5091a2987d1d9d1fc2647e8R1-R4))
*  Add references to state law for main income tax brackets and rates for each filing status ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-2889691fd9a3a2bf6ed7b39a98079c6213f0f653fe886e19d089dd7de535c87aR10-R11), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-bfdc8e488fea6cc2546b5d439b7d376c2b069168de50e8e4da404afd98efd536R10-R11), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-64df3d27642bdee378ab2bafaab739f1d98a2056759d917e297ea803523e2886R10-R11), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-b9922246939c789ffc3dadb8f86a3d1fe96efeb8bed29ce99c48906945a77f05R10-R11), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-7d8a0ea36cda46a46a972bc08c8728f5e01b32e188ec510966ba6d2c848aee10R10-R11))
*  Update main income tax brackets and rates for each filing status for 2022 ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-2889691fd9a3a2bf6ed7b39a98079c6213f0f653fe886e19d089dd7de535c87aL27-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-bfdc8e488fea6cc2546b5d439b7d376c2b069168de50e8e4da404afd98efd536L27-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-64df3d27642bdee378ab2bafaab739f1d98a2056759d917e297ea803523e2886L27-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-b9922246939c789ffc3dadb8f86a3d1fe96efeb8bed29ce99c48906945a77f05L27-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-7d8a0ea36cda46a46a972bc08c8728f5e01b32e188ec510966ba6d2c848aee10L27-R39))
*  Fix typos in tax rates for first bracket for joint and widow filing statuses for main income tax ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-bfdc8e488fea6cc2546b5d439b7d376c2b069168de50e8e4da404afd98efd536L18-R20), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-7d8a0ea36cda46a46a972bc08c8728f5e01b32e188ec510966ba6d2c848aee10L18-R20))
*  Remove empty line at the end of `single.yaml` for main income tax ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-b9922246939c789ffc3dadb8f86a3d1fe96efeb8bed29ce99c48906945a77f05L51))
*  Add new parameters for recapture base and incremental benefit amounts for each filing status for supplemental income tax, with description, metadata, reference, and brackets for 2021 and 2022 ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-8c3a350c998aa37ee07a0dfa4d9391d372b28796989d060c8c6236cf59f69a29R1-R33), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-c7b2708a89d6e672ae1e170148f152c33a38eeb1770fe720b466b9685910abacR1-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-139bd37b5b4a54fe7f9d0977b9cd440b08e65e3742c1b53b05d41ee3a5770e30R1-R32), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-2e0b816a417901f0b19a0b0276dcb80069fbf6df6ed7ad00ad5105d380e01a53R1-R32), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-ebf90a8920ec742e5729abfce295d080bd2c70f5d32b075689a92be1a0e5b46aR1-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-8cd7a16a1f24fc5aed3615a4b3daa10439f210eceda95136d51858e02ddaff5aR1-R33), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-4bc3c8ed28980eb2d477f913c10f5623216be9607020b43c67d9df35b597ff9bR1-R39), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-cd857c83ffcbcfe03a91104edfee9c8ce95786218d25c5b3a5872c091434b833R1-R33), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-65f718381fbc292f580ba6ff05cf42c5ea84ae59ea70e99309c58d0079f1c4e9R1-R33), [link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-5b9ec829447e9e552aaefcf129a274dc1d569f26b6e45d26fcd559c9ed55cd97R1-R39))
*  Rename variable `sup_tax` to `p` and add variable `ny_main_income_tax` in formula for `ny_supplemental_tax` ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-9f494b1b792e3739ee4c3985ee07d553ab214a1454e1990fb84c6944169b5566L16-R18))
*  Fix bug in formula for `ny_supplemental_tax` where previous threshold should be based on `ny_taxable_income` instead of `ny_agi` ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-9f494b1b792e3739ee4c3985ee07d553ab214a1454e1990fb84c6944169b5566L41-R42))
*  Rewrite formula for `ny_supplemental_tax` using new parameters and accounting for edge cases ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-9f494b1b792e3739ee4c3985ee07d553ab214a1454e1990fb84c6944169b5566L47-R130))
*  Add unit tests for `ny_supplemental_tax` for different filing statuses, periods, and income levels ([link](https://github.com/PolicyEngine/policyengine-us/pull/3373/files?diff=unified&w=0#diff-b5bb90710a4d283883eb8a9b9c0f7d1cdb091416a6ddaeb1f72a88a11f7d7eaaR1-R107))

### CODE REVIEW

File: changelog_entry.yaml
Comment (PavelMakarchuk) [inline changelog_entry.yaml:1]: need changelog

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+description: NY supplemental tax incremental benefit amount for a head of household filer. 
+metadata:
+  type: single_amount
+  threshold_unit: currency-USD
+  amount_unit: currency-USD
+  label: NY supplemental tax incremental benefit amount for head of household filers
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:6]:
```suggestion
  label: New York supplemental tax incremental benefit head of household amount 
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:6]: similar for all incremental benefit and recapture base files

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:6]: spell out New York in all labels and descriptions

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+  reference:
+    - title: Section 601 Imposition of tax, (d-2) for 2021 and (d-3) for 2022
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:8]: is there a tax form reference as well (2021 and 2022)? - all yaml files

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
+description: NY supplemental tax incremental benefit amount for a head of household filer.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:1]:
```suggestion
description: New York supplemental tax incremental benefit amount for a head of household filers. 
```

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:1]: similar for all incremental benefit and recapture base files

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/max_agi.yaml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+description: Filers with AGI below this amount pay the NY supplemental tax.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ny/tax/income/supplemental/max_agi.yaml:1]:
```suggestion
description: New York requires filers with adjusted gross income below this amount to pay the supplemental tax.
```

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/max_agi.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ny/tax/income/supplemental/max_agi.yaml:1]: rename to agi_limit

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/max_agi.yaml [hunk 5]
```diff
@@ -0,0 +1,10 @@
...
+values:
+  2021-01-01: 25_000_000
+metadata:
+  unit: currency-USD
+  period: year
+  label: NY State supplemental tax maximum AGI
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ny/tax/income/supplemental/max_agi.yaml:7]:
```suggestion
  label: New York supplemental tax adjusted gross income cap
```

File: policyengine_us/variables/gov/states/ny/tax/income/ny_supplemental_tax.py [hunk 6]
```diff
@@ -52,16 +53,101 @@ def formula(tax_unit, period, parameters):
...
+        #             max_(sup_tax.min_agi + 1, ny_taxable_income)
+        #         )
+        #         for scale in scales
+        #     ],
+        # )
 
-        target_tax = rate * ny_taxable_income
-        difference = target_tax - tax_unit("ny_main_income_tax", period)
-        return phase_in_fraction * difference
+        # target_tax = rate * ny_taxable_income
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ny/tax/income/ny_supplemental_tax.py:151]: delete code if not needed

**Review by PavelMakarchuk — CHANGES_REQUESTED** [review]

File: policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml (hunk 2 above)
Comment (Shinuing) [reply policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:8]: 2021 added. Let me know if you would like me to check the calculations for 2022 as the NY main income tax rates will need to be changed as well.

Comment (PavelMakarchuk) [reply policyengine_us/parameters/gov/states/ny/tax/income/supplemental/incremental_benefit/head_of_household.yaml:8]: That would be great! Could you add the 2022 rates to the parameters here as well

**Comment by Shinuing** [comment]
Fixed the supplemental tax calculation, but there are still two errors @PavelMakarchuk 

1. `ny_income_tax `cannot be within the absolute margin of error 
2. no idea how the `ny_income_tax_before_credits` test was derived, the calculation doesn't match the variable

**Comment by PavelMakarchuk** [comment]
> Fixed the supplemental tax calculation, but there are still two errors @PavelMakarchuk
> 
> 1. `ny_income_tax `cannot be within the absolute margin of error
> 2. no idea how the `ny_income_tax_before_credits` test was derived, the calculation doesn't match the variable

@Shinuing 
1. Is it a rounding or a calculation error

2. `ny_income_tax_before_credits` should be rename to `ny_income_tax_before_non_refundable_credits` and since it is just an adds function, it should not have a unit test. 
But is adding the main and supplemental income tax correct?

**Review by PavelMakarchuk — APPROVED** [review]

**Review by MaxGhenis — APPROVED** [review]
passes tests so let's merge
